### DISCUSSION TITLE
Does `ParseHostURL()` work correctly?

### DISCUSSION THREAD

**Discussion by matejvasek** [moby/moby#46015] [General]
```go
func TestParseHost(t *testing.T) {
	u, err := client.ParseHostURL("http://localhost/docker/api/subpath/")
	if err != nil {
		t.Fatal(err)
	}
	t.Logf("%#v", *u)
}
```
output is:
```
url.URL{Scheme:"http", Opaque:"", User:(*url.Userinfo)(nil), Host:"localhost/docker/api/subpath/", Path:"", RawPath:"", OmitHost:false, ForceQuery:false, RawQuery:"", Fragment:"", RawFragment:""}
```
I don't think this is correct.
IMO `Host` should be `localhost` and `Path` should be `/docker/api/subpath/`.

**Comment by matejvasek** [comment]
The thing is with go 1.20.6 an issue appeared (possible related to https://github.com/golang/go/issues/61076).
If I do something like `client.WithHost("http://placeholder/")` then
I will get ` error during connect: Get "http://placeholder%2F/_ping` not that `%2F` which is encoded slash.
I believe this is caused by `ParseHostURL()` not working properly.

It looks like proper splitting of host/addres and path is done for [TCP](https://github.com/moby/moby/blob/8e51b8b59cb81efba9491bb98a80396f2e578fe4/client/client.go#L331-L338).
But it makes no sense to me, if anything it should be done for HTTP(S) also.

**Reply by thaJeztah** [reply]
The Go issue has been resolved (but not yet in a tagged release) through this PR (and various related ones, due to circular dependencies):

- https://github.com/moby/moby/pull/45942

And further details / discussion on the associated ticket;

- https://github.com/moby/moby/issues/45935

As to `ParseHostURL` working correctly; it works "correctly" (in some interpretations of "correctly"). First and foremost; don't use it as a generic utility (we should probably add a big disclaimer to that function's doc).

The CLI itself does not accept `http(s):` as host, only `tcp`, `unix`, `ssh` and `npipe`, but validation for this happens before we hit `ParseHostURL`;

```bash
docker -H http://localhost:2375 ps
Failed to initialize: unable to resolve docker endpoint: invalid bind address format: http://localhost:2375

docker -H tcp://localhost:2375 ps
CONTAINER ID   IMAGE     COMMAND   CREATED   STATUS    PORTS     NAMES
```

W.R.T. 'in some interpretations of "correctly"';

There's a lot of legacy code in this area, and some rather "curious" choices made in the past, not all of them well-documented (early days of Docker, things moved fast). Some of those choices were to work around limitations of Go at the time (some code dates back to when Go was at v0.7 or even older), and such code / logic has found its way into many parts of the code (and external consumers using the code). There's a lot of cleaning up / refactoring to be done to address that, but as it affects a lot of the code, and affects some very "core" components, we need to do so granular, so for this specific issue, we opted to only address the issue at hand, and otherwise "if it ain't broke, don't fix it".

There may also be some follow-up discussions to be had with the Go maintainers, as Go stdlib is not strictly complying to the RFCs (https://github.com/moby/moby/pull/45942#discussion_r1261473862), and requires a host to be set (even though it unused), which is what ultimately led to this issue (just set any value to make sure it's not-empty) https://pkg.go.dev/github.com/docker/docker@v24.0.5-0.20230718221249-d4a26c153000+incompatible/client#DummyHost
